crossing-training.lic

```diff
@@ -382,27 +382,51 @@ class CrossingTraining
   def calculate_mana(min, more, discern_data, cyclic)
     total = min + more
     total = (total * @settings.prep_scaling_factor).floor
-    discern_data['mana'] = [(total / 4.0).ceil, min].max
+    discern_data['mana'] = [(total / 5.0).ceil, min].max
     remaining = total - discern_data['mana']
-    if remaining > @settings.cambrinth_cap
-      discern_data['mana'] = discern_data['mana'] + (remaining - @settings.cambrinth_cap)
+    if @settings.cambrinth_items[0]['name']
+      cambrinth_items = @settings.cambrinth_items
+    else
+      cambrinth_items = []
+      cambrinth_items[0] = {}
+      cambrinth_items[0]['name'], cambrinth_items[0]['cap'], cambrinth_items[0]['stored'] = map_cambrinth(@settings)
+    end
+    total_cambrinth_cap = 0
+    cambrinth_items.each do |item|
+      total_cambrinth_cap += item['cap']
+    end
+    total_cambrinth_charges = 0
+    charges_count_floor = 4
+    cambrinth_items.each do |item|
+      item['charges'] = ((item['cap'].to_f / total_cambrinth_cap) * charges_count_floor).ceil
+      total_cambrinth_charges += item['charges']
+    end
+    if remaining > total_cambrinth_cap
+      discern_data['mana'] = discern_data['mana'] + (remaining - total_cambrinth_cap)
       remaining = total - discern_data['mana']
     end
     if cyclic
       discern_data['cambrinth'] = nil
       discern_data['mana'] = discern_data['mana'] + remaining
     elsif remaining > 0
-      discern_data['cambrinth'] = []
-      step_size = (remaining / 3.0).ceil
-      while remaining > 0
-        discern_data['cambrinth'] << [remaining, step_size].min
-        remaining -= step_size
+      total_cambrinth_mana = [remaining, total_cambrinth_cap].min
+      cambrinth_items.each_with_index do |item, index|
+        discern_data['cambrinth'] ||= []
+        charge_amount = (total_cambrinth_mana / total_cambrinth_charges) * item['charges']
+        discern_data['cambrinth'][index] = []
+        charge_amount.times do |i|
+          discern_data['cambrinth'][index][i%item['charges']] += 1
```

Comment:
Has this been tested? Unless lich has messed with nil again  this seems like it will fail because you're trying to add 1 to nil (an undefined array entry).

Reply:
It worked when testing before.  I just tried it again with some echos added before and after assigning values to the array to make sure it was assigning as intended:

```
        discern_data['cambrinth'][index] = []
        echo "discern_data['cambrinth'][index] for #{item['name']} is equal to #{discern_data['cambrinth'][index]}"
        charge_amount.times do |i|
          discern_data['cambrinth'][index][i%item['charges']] += 1
        end
        echo "discern_data['cambrinth'][index] for #{item['name']} is equal to #{discern_data['cambrinth'][index]}"
```

And I got this when running it in game:
```
[crossing-training: discern_data['cambrinth'][index] for cambrinth armband is equal to []]
[crossing-training: discern_data['cambrinth'][index] for cambrinth armband is equal to [9, 9, 9]]
[crossing-training: discern_data['cambrinth'][index] for cambrinth hairnet is equal to []]
[crossing-training: discern_data['cambrinth'][index] for cambrinth hairnet is equal to [9]]
[crossing-training: discern_data['cambrinth'][index] for cambrinth moonflower is equal to []]
[crossing-training: discern_data['cambrinth'][index] for cambrinth moonflower is equal to [9, 9]]
```

Reply:
Yeah it's just lich being dumb and redefining nil + value to be value instead of an error.

Reply:
Would it make more sense to add
`discern_data['cambrinth'][index][i%item['charges']] ||= 0`
before
`discern_data['cambrinth'][index][i%item['charges']] += 1`
?
That way it will explicitly assign it a value of 0 if it's nil.